Idea box: Interactive maps with population changes from 2014 to 2020, for US cities/counties 
Compare these to Missoula City/County. 
Also potential interactive maps representing county budget cuts? Demographics? 

https://www.visualnoise.io/interactive-maps-with-python-pandas-and-plotly/

Mapping out US Census Data: https://towardsdatascience.com/mapping-us-census-data-with-python-607df3de4b9c


Clean up: remove all repeat imports (for example import pandas as pd)

Missoula Metrics

Population Estimates
Total Jurisdiction Operating Budget: 
Total FTEs (Full Time Employee Equivalents) 
Acres of park maintained
Number of parks maintained 

Calculate: Acres of park per 1000 residents


Cluster Analysis: 2022 





Population

In [33]:
#reference: https://www.census.gov/geographies/reference-files/2010/geo/state-area.html.
import censusdata

# Define the year and city you want to pull data for
year = 2017
city = "Missoula city, Montana"

# Define the population data variable
population_data = censusdata.download("acs5", year, censusdata.censusgeo([('state', '30'), ('place', '50200')]), ['B01003_001E'])

# Print the population data for the specified city and year
print(f"Population of {city} in {year}: {population_data['B01003_001E'][0]}")



Population of Missoula city, Montana in 2017: 70847


In [60]:
#Percent change in population from 2000 to 2020 
year1=2000
year2=2020
percent_change=round(((74994-57871)/74994)*100)

#print (f"Percent change of percent_change)
print(f"Percent change of {city} from {year1} to {year2}  : {percent_change}")



Percent change of Missoula city, Montana from 2000 to 2020  : 23


Population by year: 

2000- 57871 

2017-70847

2018-72125

2019-73710

2020-74994

Wage-Based Gini

Wage-based Gini of Missoula city, Montana in 2017: 0.4867

Wage-based Gini of Missoula city, Montana in 2018: 0.485

Wage-based Gini of Missoula city, Montana in 2019: 0.4866

Wage-based Gini of Missoula city, Montana in 2020: 0.4862



In [5]:
#reference: https://www.census.gov/geographies/reference-files/2010/geo/state-area.html.
import censusdata

# Define the year and city you want to pull data for
year = 2020
city = "Missoula city, Montana"

# Define the population data variable
gini_data = censusdata.download("acs5", year, censusdata.censusgeo([('state', '30'), ('place', '50200')]), ['B19083_001E'])

# Print the population data for the specified city and year
print(f"Wage-based Gini of {city} in {year}: {gini_data['B19083_001E'][0]}")

Wage-based Gini of Missoula city, Montana in 2020: 0.4862


Median Household Income (B19013_001E): MEDIAN HOUSEHOLD INCOME THE PAST 12 MONTHS (IN 2019 INFLATION-ADJUSTED DOLLARS)

Median Household Income of Missoula city, Montana in 2020: 50947

Median Household Income of Missoula city, Montana in 2019: 47426

Median Household Income of Missoula city, Montana in 2018: 45010

Median Household Income of Missoula city, Montana in 2017: 43602



In [15]:
year = 2017
city = "Missoula city, Montana"

# Define the population data variable
median_income = censusdata.download("acs5", year, censusdata.censusgeo([('state', '30'), ('place', '50200')]), ['B19013_001E'])

# Print the population data for the specified city and year
print(f"Median Household Income of {city} in {year}: {median_income['B19013_001E'][0]}")

Median Household Income of Missoula city, Montana in 2017: 39677


Median Monthly Housing Costs (B25105_001E) MEDIAN MONTHLY HOUSING COSTS (DOLLARS)

2017 Median Monthly Housing Costs of Missoula city, Montana in 2017: 867
2018 Median Monthly Housing Costs of Missoula city, Montana in 2018: 888
2019 Median Monthly Housing Costs of Missoula city, Montana in 2019: 932
2020 Median Monthly Housing Costs of Missoula city, Montana in 2020: 996

In [19]:
year = 2020
city = "Missoula city, Montana"

# Define the population data variable
monthly_costs = censusdata.download("acs5", year, censusdata.censusgeo([('state', '30'), ('place', '50200')]), ['B25105_001E'])

# Print the population data for the specified city and year
print(f"Median Monthly Housing Costs of {city} in {year}: {monthly_costs['B25105_001E'][0]}")

Median Monthly Housing Costs of Missoula city, Montana in 2020: 996


Employment Status (B23025_004E) EMPLOYMENT STATUS FOR THE POPULATION 16 YEARS AND OVER

2017: Employment Status of Missoula city, Montana in 2017: 39677
2018: Employment Status of Missoula city, Montana in 2018: 41702
2019: Employment Status of Missoula city, Montana in 2019: 42856
2020: Employment Status of Missoula city, Montana in 2020: 44185

In [25]:
year = 2017
city = "Missoula city, Montana"

# Define the population data variable
employment_status = censusdata.download("acs5", year, censusdata.censusgeo([('state', '30'), ('place', '50200')]), ['B23025_004E'])

# Print the population data for the specified city and year
print(f"Employment Status of {city} in {year}: {employment_status['B23025_004E'][0]}")

Employment Status of Missoula city, Montana in 2017: 39677


Unemployment Status(B23025_005E) : EMPLOYMENT STATUS FOR THE POPULATION 16 YEARS AND OVER

2017: Unemployment Status of Missoula city, Montana in 2017: 2637
2018: Unemployment Status of Missoula city, Montana in 2018: 2029
2019: Unemployment Status of Missoula city, Montana in 2019: 2208
2020: Unemployment Status of Missoula city, Montana in 2020: 2336


In [31]:
year = 2020
city = "Missoula city, Montana"

# Define the population data variable
unemployment_status = censusdata.download("acs5", year, censusdata.censusgeo([('state', '30'), ('place', '50200')]), ['B23025_005E'])

# Print the population data for the specified city and year
print(f"Unemployment Status of {city} in {year}: {unemployment_status['B23025_005E'][0]}")

Unemployment Status of Missoula city, Montana in 2020: 2336


In [91]:
import pandas as pd
from census import Census
from us import states
from sklearn.cluster import KMeans


In [72]:
import censusdata 
import pandas as pd
# Define variables as dictionaries
variable1 = {'name': 'B23025_004E', 'label': 'Employed'}
variable2 = {'name': 'B23025_005E', 'label': 'Unemployed'}
variable3 = {'name': 'B01003_001E', 'label': 'Population'}
variable4 = {'name': 'B25105_001E', 'label': 'MonthlyCosts'}
variable5 = {'name': 'B19013_001E', 'label': 'Income'}


# Define other necessary parameters
year = 2019
state = '30'
place = '50200'

# Download data for each variable and combine into a single DataFrame
data1 = censusdata.download("acs5", year, censusdata.censusgeo([('state', state), ('place', place)]), [variable1['name']])
data2 = censusdata.download("acs5", year, censusdata.censusgeo([('state', state), ('place', place)]), [variable2['name']])
data3 = censusdata.download("acs5", year, censusdata.censusgeo([('state', state), ('place', place)]), [variable3['name']])
data4 = censusdata.download("acs5", year, censusdata.censusgeo([('state', state), ('place', place)]), [variable4['name']])
data5 = censusdata.download("acs5", year, censusdata.censusgeo([('state', state), ('place', place)]), [variable5['name']])

# Rename columns to include variable labels
data1.rename(columns={variable1['name']: variable1['label']}, inplace=True)
data2.rename(columns={variable2['name']: variable2['label']}, inplace=True)
data3.rename(columns={variable3['name']: variable3['label']}, inplace=True)
data4.rename(columns={variable4['name']: variable4['label']}, inplace=True)
data5.rename(columns={variable5['name']: variable5['label']}, inplace=True)

df = pd.concat([data1, data2, data3, data4, data5], axis=1)

# Export the dataframe to a CSV file
#df.to_csv('output.csv', index=False)

# Merge the  DataFrames
#result = data1.merge(data2, left_index=True, right_index=True)
#merged_data = data1.merge(data2, on='year').merge(data3, on='year').merge(data4, on='year').merge(data5, on='year')

# Merge data frames
#merged_data = data1.merge(data2, on='50200').merge(data3, on='50200').merge(data4, on='50200').merge(data5, on='50200')

# Save merged data as CSV file
#merged_data.to_csv('merged_data.csv', index=False)

# Export the dataframe to a CSV file
df.to_csv('census_output.csv', index=False)


# Print confirmation message
print('Output file created successfully.')


Output file created successfully.


In [ ]:
year = 2020
city = "Missoula city, Montana"

# Define the population data variable
employment_status = censusdata.download("acs5", year, censusdata.censusgeo([('state', '30'), ('place', '50200')]), ['B23025_004E'])

# Print the population data for the specified city and year
print(f"Employment Status of {city} in {year}: {employment_status['B23025_004E'][0]}")

In [ ]:
import pandas as pd

census_data = pd.read_csv("census_data.csv")

In [73]:
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

In [170]:
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

# Load the census data
#census_data = censusdata.download

year = 2020
city = "Missoula city, Montana"

# Define the population data variable
unemployment_status = censusdata.download("acs5", year, censusdata.censusgeo([('state', '30'), ('place', '50200')]), ['B23025_005E'])

# Print the population data for the specified city and year
print(f"Unemployment Status of {city} in {year}: {unemployment_status['B23025_005E'][0]}")

# Preprocess the data
features = ['Median Age', 'Per Capita Income', 'Poverty Rate']
X = census_data[features].values
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Find the optimal number of clusters using the elbow method
wcss = []
for i in range(1, 11):
    kmeans = KMeans(n_clusters=i, init='k-means++', random_state=42)
    kmeans.fit(X_scaled)
    wcss.append(kmeans.inertia_)
plt.plot(range(1, 11), wcss)
plt.title('Elbow Method')
plt.xlabel('Number of Clusters')
plt.ylabel('WCSS')
plt.show()

# Fit the KMeans model
n_clusters = 4
kmeans = KMeans(n_clusters=n_clusters, init='k-means++', random_state=42)
kmeans.fit(X_scaled)

# Add the cluster labels to the census data
census_data['Cluster'] = kmeans.labels_

# Visualize the results
for cluster in range(n_clusters):
    cluster_data = census_data[census_data['Cluster'] == cluster]
    plt.scatter(cluster_data['Median Age'], cluster_data['Per Capita Income'], label=f'Cluster {cluster}')
plt.xlabel('Median Age')
plt.ylabel('Per Capita Income')
plt.legend()
plt.show()


TypeError: 'function' object is not subscriptable

Like accounts: followers, likes, comments, how often posted in last 60 days (is this even relevant) 

In [89]:
import json
import requests
import numpy as np
import pandas as pd
from tqdm.notebook import tqdm
import matplotlib.pyplot as plt
import geopandas

In [90]:
# Read file
db = geopandas.read_file("../data/sandiego/sandiego_tracts.gpkg")

DriverError: ../data/sandiego/sandiego_tracts.gpkg: No such file or directory

In [85]:
state_codes = [
    str(elem).zfill(2) 
    for elem in list(range(1, 57)) 
    if elem not in (2, 3, 7, 11, 14, 15, 43, 52)
]
print("# of State FIPS Codes:", len(state_codes))
state_codes[0:5]

# of State FIPS Codes: 48


['01', '04', '05', '06', '08']

In [129]:
import pandas as pd

try:
    df = pd.read_csv('/Users/krusty/Desktop/Capstone/NRPADATAALL2.csv')
except pd.errors.EmptyDataError:
    print('Error: No data found in file.')


Error: No data found in file.


In [130]:
import pandas as pd

# Replace the file path with your own CSV file path
file_path = "/Users/krusty/Desktop/Capstone/NRPADATAALL2.csv"

# Read the CSV file into a Pandas dataframe
data = pd.read_csv(file_path, header=None)

# Do something with the data
print(data.head())


EmptyDataError: No columns to parse from file

In [135]:
import pandas as pd

# Read Excel file
df = pd.read_excel('/Users/krusty/Desktop/Capstone/NRPADATAALL2.xlsx', sheet_name='NRPADATAALL2')

# Print the dataframe
# print(df.head())

#print all columns 
for col in df.columns:
    print(col)







Agency
Year
Region
Country
Operating expenditures per capita
Revenue per capita
Total revenue to total operating expenditures
Total tax expenditures per capita
Park operating expenditures per acre of parkland
Operating expenditures per acre of parkland
Operating expenditures per acres of parks and non-park sites
Operating expenditures per FTE
FTE's per 10,000 population
Acres of parks per 1,000 residents
Number of residents per park
Number of acres per park
Number of participants per program
Ratio of fee programs to all programs
Ratio of building attendance to park attendance
 Jurisdiction Information
Which of the following best describes your agency/department's jurisdiction type?
If 'Other,' please describe:
Current Year Jurisdiction total operating budget
Current Year Jurisdiction annual capital budget
Square mileage of incorporated jurisdiction
Population
In what country is your agency located?
Jurisdiction State/Province
What were the total operating expenditures for your agency d

In [162]:
# example data
data = [
    ['Agency', 'Year', 'Region', 'Current Year Jurisdiction total operating budget', 'Population', 'Total number of parks', 'Total number of non-park sites', 'Total number of FTEs'],
    ['Mansfield Parks and Recreation Department', '2016', 'TX', '$4,775,144', '67628.0', '17', '5', '50'],
    ['Another Agency', '2021', 'MA', '$200,000', '20,000', '75', '30', '10'],
]



# set the index of the column to be quoted
quote_index = 0

# write CSV file with quotes around selected column
with open('output2.csv', 'w', newline='') as f:
    writer = csv.writer(f)
    for row in data:
        # add quotes around the selected column
        row[quote_index] = '"' + row[quote_index] + '"'
        writer.writerow(row)

In [167]:
import csv

data = [
    ['Agency', 'Year', 'Region', 'Current Year Jurisdiction total operating budget', 'Population', 'Total number of parks', 'Total number of non-park sites', 'Total number of FTEs'],
    ['Mansfield Parks and Recreation Department', '2016', 'TX', '4,775,144', '67628.0', '17', '5', '50'],
    ['Another Agency', '2021', 'MA', '200,000', '20,000', '75', '30', '10'],
]

# set the index of the column to be quoted
quote_index = 0

# write CSV file with quotes around selected column and convert budget to float
with open('output2.csv', 'w', newline='') as f:
    writer = csv.writer(f)
    for row in data:
        # add quotes around the selected column
        row[quote_index] = '"' + row[quote_index] + '"'
        
        
        writer.writerow(row)


In [74]:
#Cluster Analysis 
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import csv

In [81]:


# # Load the data from a CSV file
data = pd.read_csv('census_output.csv')

# Prepare the data for clustering
X = data.values
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Find the optimal number of clusters using the elbow method
wcss = []
for i in range(1, 11):
    kmeans = KMeans(n_clusters=i, init='k-means++', max_iter=300, n_init=10, random_state=0)
    kmeans.fit(X_scaled)
    wcss.append(kmeans.inertia_)

# Plot the elbow curve
plt.plot(range(1, 11), wcss)
plt.title('Elbow Method')
plt.xlabel('Number of clusters')
plt.ylabel('WCSS')
plt.show()






# # Find the optimal number of clusters using the elbow method
# wcss = []
# for i in range(1, 11):
#     kmeans = KMeans(n_clusters=i, init='k-means++', max_iter=300, n_init=10, random_state=0)
#     kmeans.fit(X_scaled)
#     wcss.append(kmeans.inertia_)
#     plt.plot(range(1, 11), wcss)
#     plt.title('Elbow Method')
#     plt.xlabel('Number of clusters')
#     plt.ylabel('WCSS')
#     plt.show()

# # Perform the clustering
# k = 3  # Replace with the optimal number of clusters you found
# kmeans = KMeans(n_clusters=k, init='k-means++', max_iter=300, n_init=10, random_state=0)
# y_kmeans = kmeans.fit_predict(X_scaled)

# # Add the cluster labels to the original data
# data['Cluster'] = y_kmeans

# # Write the clustered data to a new CSV file
# data.to_csv('clustered_data.csv', index=False)  # Replace with the name of the output file you want to create


ValueError: n_samples=1 should be >= n_clusters=2.

In [156]:
# import pandas as pd

# df = pd.read_csv('/Users/krusty/Desktop/Capstone/NRPADATAALL2.csv')

# # Print the column names of the DataFrame
# #print(df.columns.tolist())

# # Select certain columns
# selected_columns = ['Agency', 'Year', 'Region', 'Current Year Jurisdiction total operating budget', 'Population', 'Total number of parks', 'Total number of non-park sites', 'Total number of FTEs', ]  # Replace with the column names you want to select
# # data = 'Stoughton Parks & Recreation Department','2011','WI','$11,067,754','12893.0','20','0,8'


# # example data
# # data = [
# #     ['Agency', 'Year', 'Region', 'Current Year Jurisdiction total operating budget', 'Population', 'Total number of parks', 'Total number of non-park sites', 'Total number of FTEs'],
# #     ['Stoughton Parks & Recreation Department', '2022', 'Northeast', '$100,000', '10,000', '50', '20', '5'],
# #     ['Another Agency', '2022', 'Southwest', '$200,000', '20,000', '75', '30', '10'],
# # ]




# df_selected = df[selected_columns]

# # Write the selected columns to a new CSV file
# df_selected.to_csv('output.csv', index=False)  # Replace with the name of the output file you want to create


# # Define the lower and upper bounds of the population range
# lower_bound = 50000
# upper_bound = 99999

# # # Filter the DataFrame to show only the rows where population falls within the range
# filtered_df = df[(df['Population'] >= lower_bound) & (df['Population'] <= upper_bound)]


# # # Display the filtered DataFrame
# print(filtered_df)
# filtered_df.to_csv('filtered.csv', index=False)




                                        Agency  Year Region        Country  \
62               Hoffman Estates Park District  2015     IL  United States   
63               Hoffman Estates Park District  2017     IL  United States   
64               Hoffman Estates Park District  2016     IL  United States   
65               Hoffman Estates Park District  2019     IL  United States   
66               Hoffman Estates Park District  2014     IL  United States   
...                                        ...   ...    ...            ...   
7851  Coeur d'Alene (City of) Parks Department  2021     ID  United States   
7852  Coeur d'Alene (City of) Parks Department  2022     ID  United States   
7907      Centerville-Washington Park District  2022     OH  United States   
7915                                      RHAA  2022     CA  United States   
7921                           City of Hanford  2022     CA  United States   

     Operating expenditures per capita Revenue per capita  \
62

In [171]:
import pandas as pd

selected_columns = ['Agency', 'Year', 'Region', 'Current Year Jurisdiction total operating budget', 
                    'Population', 'Total number of parks', 'Total number of non-park sites', 'Total number of FTEs']

df = pd.read_csv('/Users/krusty/Desktop/Capstone/NRPADATAALL2.csv')

df_selected = df[selected_columns]

df_selected.to_csv('output.csv', index=False)  # Replace with the name of the output file you want to create

lower_bound = 50000
upper_bound = 99999

filtered_df = df_selected[(df_selected['Population'] >= lower_bound) & (df_selected['Population'] <= upper_bound)]

filtered_df.to_csv('filtered.csv', index=False)


In [ ]:
# import pandas as pd

# df = pd.read_csv('/Users/krusty/Desktop/Capstone/NRPADATAALL2.csv')

# # Print the column names of the DataFrame
# #print(df.columns.tolist())

# # Select certain columns
# #selected_columns = ['Column1', 'Column3', 'Column5']  # Replace with the column names you want to select
# #df_selected = df[selected_columns]

# # Write the selected columns to a new CSV file
# #df_selected.to_csv('output.csv', index=False)  # Replace with the name of the output file you want to create


# # Define the lower and upper bounds of the population range
# lower_bound = 50000
# upper_bound = 99999

# # # Filter the DataFrame to show only the rows where population falls within the range
# filtered_df = df[(df['Population'] >= lower_bound) & (df['Population'] <= upper_bound)]

# # # Display the filtered DataFrame
# print(filtered_df)

## City of Missoula Last Five Year Population Census Estimates 

In [88]:
import requests

# Define the URL for the REST API endpoint
url = "https://services.arcgis.com/SXbDpmb7xQkk44JV/arcgis/rest/services/service_08988205ac2a4ef2ab27244db2dad485/FeatureServer/0/query"

# Define the query parameters
params = {
    "f": "geojson",
    "where": "urban_ag_review='yes'",
    "returnGeometry": "true",
    "spatialRel": "esriSpatialRelIntersects",
    "outFields": "*",
    "maxRecordCountFactor": "5",
    "outSR": "102100",
    "resultOffset": "0",
    "resultRecordCount": "5000",
    "cacheHint": "true",
    "quantizationParameters": '{"mode":"view","originPosition":"upperLeft","tolerance":1.0583354500042335,"extent":{"xmin":-180,"ymin":-90,"xmax":180,"ymax":90,"spatialReference":{"wkid":4326,"latestWkid":4326}}}'
}

# Send a GET request to the URL with the query parameters
response = requests.get(url, params=params)

# Check if the response is successful
if response.status_code == 200:
    # The response contains the data in GeoJSON format
    geojson_data = response.json()
else:
    # The response was not successful
    print(f"Error: {response.status_code} - {response.reason}")


## KNN of Cities Most Comparable to Missoula

In [ ]:
import numpy as np
from sklearn import datasets
from sklearn.neighbors import KNeighborsClassifier

# Load the iris dataset
iris = datasets.load_iris()
X = iris.data
y = iris.target

# Split the data into training and testing sets
# We will use the first 100 samples for training and the remaining for testing
X_train, y_train = X[:100], y[:100]
X_test, y_test = X[100:], y[100:]

# Create a KNN classifier with k=3 (3 nearest neighbors)
knn = KNeighborsClassifier(n_neighbors=3)

# Train the classifier on the training data
knn.fit(X_train, y_train)

# Predict the classes of the samples in the testing set
predictions = knn.predict(X_test)

# Calculate the accuracy of the classifier
accuracy = np.mean(predictions == y_test)
print("Accuracy:", accuracy)


Explanation: 

KNN classifies the new data points based on the similarity measure of the earlier stored data points. For example, if we have a dataset of tomatoes and bananas. KNN will store similar measures like shape and color. When a new object comes it will check its similarity with the color (red or yellow) and shape.

In the code block above, we split data into training and testing sets, using the first 100 samples for training and the remaining for testing. This is a common method for evaluating the performance of machine learning algorithms.

Next, we create a KNN classifier using the KNeighborsClassifier class from the sklearn.neighbors module. We specify that we want to use n_neighbors=3, meaning that the classifier will use the 3 nearest neighbors when making predictions.

We then train the classifier on the training data using the fit method, and use the predict method to make predictions on the samples in the testing set. Finally, we calculate the accuracy of the classifier by comparing the predicted classes to the true classes of the samples in the testing set.



Metrics: top categories: Budget size, Number of Parks or Sites Maintained, FTEs, Jurisdiction Population, Jurisdiction Population Per Sq Mile 

In [92]:
#trying to extract info from pdf, having issues with following two code chunks  
# extract_doc_info.py

from PyPDF2 import metadata

pdf_path = 'https://parkserve.tpl.org/mapping/historic/2022_ParkScoreRank.pdf'
   
def extract_information(pdf_path):
    with open(pdf_path,'rb') as f:
        pdf = metadata(f)
        information = pdf.getDocumentInfo()
        number_of_pages = pdf.getNumPages()

    txt = f"""
    Information about {pdf_path}: 

    Author: {information.author}
    Creator: {information.creator}
    Producer: {information.producer}
    Subject: {information.subject}
    Title: {information.title}
    Number of pages: {number_of_pages}
    """

    print(txt)
    return information

if __name__ == '__main__':
    path = 'https://parkserve.tpl.org/mapping/historic/2022_ParkScoreRank.pdf'
    extract_information(path)

ImportError: cannot import name 'metadata' from 'PyPDF2' (/Users/krusty/opt/anaconda3/lib/python3.9/site-packages/PyPDF2/__init__.py)

In [3]:
def extract_text_from_pdf(pdf_path):
    with open(pdf_path, 'rb') as f:
        #PyPDF2 = (pdf_path)
        #pdf = PyPDF2.PdfFileReader(f)
        pdf = pdf_path(f)
        num_pages = pdf.getNumPages()
        text = ''
        for i in range(num_pages):
            page = pdf.getPage(i)
            text += page.extractText()
        return text

pdf_file_path = '/Users/krusty/Desktop/Capstone/ProvoPNR.pdf'
text = extract_text_from_pdf(pdf_file_path)
print(text)

TypeError: 'str' object is not callable

## NRPA Percentile Plot

Explanation: 

In this code, we first load the data from the CSV file using the pandas.read_csv() function. This function reads the CSV file and returns a DataFrame object. We then compute the percentiles of a specific column using the quantile() method. This method computes the 25th, 50th, and 75th percentiles of the specified column.

We then create a box plot of the column using the matplotlib.pyplot.boxplot() function. This function creates a box-and-whisker plot of the column. We add horizontal lines for the percentiles using the matplotlib.pyplot.axhline() function, which adds a horizontal line to the plot at the specified y-value.

Finally, we customize the plot by setting the plot title, x-axis label, and y-axis label using the matplotlib.pyplot.title(), matplotlib.pyplot.xlabel(), and matplotlib.pyplot.ylabel() functions, respectively.

Note that the path/to/csvfile.csv in the code should be replaced with the actual path to the CSV file on your computer, and the column_name should be replaced with the name of the column you want to compute the percentiles for. Additionally, you can customize the plot further by changing the colors, linestyles, or other properties of the plot.

In [161]:
import pandas as pd
import matplotlib.pyplot as plt

# Load the data from the CSV file
data = pd.read_csv('filtered.csv')

# Compute the percentiles
percentiles = data['Current year'].quantile([0.25, 0.5, 0.75])

# Create a box plot
plt.boxplot(data['Current year'])

# Add horizontal lines for the percentiles
plt.axhline(percentiles[0.25], color='r', linestyle='--')
plt.axhline(percentiles[0.5], color='g', linestyle='--')
plt.axhline(percentiles[0.75], color='b', linestyle='--')

# Customize the plot
plt.title('Percentiles')
plt.xlabel('X label')
plt.ylabel('Y label')

# Display the plot
plt.show()


KeyError: 'Current year'

## K-Means Cluster Analysis 

In [25]:
from sklearn.cluster import KMeans
import numpy as np

# Generate sample data
np.random.seed(0)
data = np.random.rand(100, 2)

# Fit the model
kmeans = KMeans(n_clusters=3, random_state=0).fit(data)

# Predict the cluster labels for each data point
labels = kmeans.predict(data)

# View the cluster centers
cluster_centers = kmeans.cluster_centers_
print(cluster_centers)
#

[[0.53191784 0.17092862]
 [0.22477318 0.67757842]
 [0.7591731  0.6383279 ]]


#nps data: https://www.nps.gov/subjects/gisandmapping/tools-and-data.htm
#nps APIs https://www.nps.gov/subjects/developer/api-documentation.htm
#CDC EJI https://eji.cdc.gov/launcher.html


In [158]:
#Cluster analysis from NRPA Data https://nrpaparkmetrics.com/NRPA/Reports/ERReports/APRT.aspx

import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

# Load the data
data = pd.read_csv('filtered.csv')

# Preprocess the data
X = data[['Region', 'Year']].values

# Fit the model
kmeans = KMeans(n_clusters=3, random_state=0).fit(X)

# Predict the cluster labels for each data point
labels = kmeans.predict(X)

# View the cluster centers
cluster_centers = kmeans.cluster_centers_
print(cluster_centers)

# Visualize the results
plt.scatter(X[:, 0], X[:, 1], c=labels)
plt.scatter(cluster_centers[:, 0], cluster_centers[:, 1], marker='x', s=200, linewidths=3, color='r')
plt.xlabel('Region')
plt.ylabel('Year')
plt.show()

#top categories: Budget size, Number of Parks or Sites Maintained, FTEs, Jurisdiction Population, 
#Jurisdiction Population Per Sq Mile 

ValueError: could not convert string to float: 'IL'

To plot a geoid in Python from a CSV file, you will need to first convert the CSV file to a GeoDataFrame, and then use the plot() method to create a map of the geoid. Here's some example code to get you started:

In this code, we first load the data from the CSV file using the geopandas.read_file() function. This function reads the CSV file and returns a DataFrame object. We then convert the data to a GeoDataFrame using the gpd.GeoDataFrame() function. This function takes the DataFrame and the longitude and latitude columns as input, and creates a new GeoDataFrame with a Point geometry column.

We then plot the geoid using the GeoDataFrame.plot() method, which creates a map of the geoid. We customize the plot by setting the figure size using the matplotlib.pyplot.figure() function, setting the plot title using the matplotlib.pyplot.title() function, and turning off the axis labels using the matplotlib.pyplot.axis() function.

Finally, we display the plot using the matplotlib.pyplot.show() function.

Note that the path/to/csvfile.csv in the code should be replaced with the actual path to the CSV file on your computer. Additionally, you may need to adjust the longitude and latitude column names in the gpd.points_from_xy() function to match the column names in your CSV file.

In [1]:
import geopandas as gpd
import matplotlib.pyplot as plt

# Load the data from the CSV file
data = gpd.read_file('/Users/krusty/Desktop/Capstone/MTEJI22.csv')

# Convert the data to a GeoDataFrame
geodata = gpd.GeoDataFrame(data, geometry=gpd.points_from_xy(data.Longitude, data.Latitude))

# Plot the geoid
geodata.plot(figsize=(10,10))

# Customize the plot
plt.title('Geoid')
plt.axis('off')

# Display the plot
plt.show()


AttributeError: module 'lib' has no attribute 'OpenSSL_add_all_algorithms'

## Analyzing Comparable City Comprehensive Plans

In [19]:
from pdfminer.pdfinterp import PDFResourceManager, PDFPageInterpreter
from pdfminer.converter import TextConverter
from pdfminer.layout import LAParams
from pdfminer.pdfpage import PDFPage
from io import StringIO

def extract_text_from_pdf(pdf_path):
    with open(pdf_path, 'rb') as file:
        resource_manager = PDFResourceManager()
        string_io = StringIO()
        converter = TextConverter(resource_manager, string_io, codec='utf-8', laparams=LAParams())
        page_interpreter = PDFPageInterpreter(resource_manager, converter)
 
        for page in PDFPage.get_pages(file, caching=True, check_extractable=True):
            page_interpreter.process_page(page)
 
        text = string_io.getvalue()
 
        # close open handles
        converter.close()
        string_io.close()
 
        return text

if __name__ == '__main__':
    print(extract_text_from_pdf('/Users/krusty/Desktop/Capstone/BPRDPlan.pdf'))
    
   

BEND PARK & RECREATION DISTRICT
COMPREHENSIVE PLAN

Due to the nature of this document, not all pages are in an accessible format. Please contact Quinn 

Keever at Quinn@bendparksandrec.org to make an accommodation request.

ADOPTED JULY 2018 

A MESSAGE 
FROM THE 
EXECUTIVE 
DIRECTOR

Abraham Lincoln once said, “Give me six hours to 
chop down a tree and I will spend the first four hours 
sharpening the axe.” 

Planning the future of your park and recreation district 
is one of the most important things we can do to assure 
that the resources we are entrusted to manage are put 
to good use.  Bend Park and Recreation District listens 
to our residents’ needs and desires and does our best 
to achieve the community’s vision for their park and 
recreation system. Listening to our residents has helped 
us develop one of the most diverse recreation programs 
in the state, and a park system envied across the nation.  
We could not do this without your input, your support, 
and your trust.


In [24]:
pdf_text = extract_text_from_pdf('/Users/krusty/Desktop/Capstone/BPRDPlan.pdf')
word = "park"
word2 = "geospatial"

#if word in pdf_text:
    #print(f"The word '{word}' was found in the text")
#else:
    #print(f"The word '{word}' was not found in the text")

count = pdf_text.count(word)
count2 = pdf_text.count(word2)
print(f"The word '{word}' was found {count} times in the text")
print(f"The word '{word2}' was found {count2} times in the text")

#Provo: Provo found 247 times in the text and climate found 3 times 
#Bend: Bend found 179 times, climate found 0, sustainable 1 time, inclusive-1, housing-5
#park-471, geospatial-9
    
    


The word 'park' was found 471 times in the text
The word 'geospatial' was found 9 times in the text


In [ ]:
from pdfminer.pdfinterp import PDFResourceManager, PDFPageInterpreter
from pdfminer.converter import TextConverter
from pdfminer.layout import LAParams
from pdfminer.pdfpage import PDFPage
from io import StringIO

def extract_text_from_pdf2(pdf_path):
    with open(pdf_path, 'rb') as file:
        resource_manager2 = PDFResourceManager()
        string_io2 = StringIO()
        converter2 = TextConverter(resource_manager, string_io, codec='utf-8', laparams=LAParams())
        page_interpreter2 = PDFPageInterpreter(resource_manager, converter)
 
        for page in PDFPage.get_pages(file, caching=True, check_extractable=True):
            page_interpreter.process_page(page)
 
        text = string_io.getvalue()
 
        # close open handles
        converter.close()
        string_io.close()
 
        return text

if __name__ == '__main__':
    print(extract_text_from_pdf('https://www.bendparksandrec.org/wp-content/uploads/2018/07/BPRD-Comp-Plan-Adopted-for-web.pdf'))
    
   

# Building Website Request to Scrape Similar Metrics

In [ ]:
import requests
from bs4 import BeautifulSoup

In [3]:
#potential: Bend, OR; Provo, UT; Bozeman, MT; Billings, MT

In [4]:
# Make a request to the website
url = "https://www.bendparksandrec.org/wp-content/uploads/2018/07/BPRD-Comp-Plan-Adopted-for-web.pdf"
response = requests.get(url)

# Check if the request was successful
if response.status_code == 200:
    # Parse the HTML content
    soup = BeautifulSoup(response.content, "html.parser")


Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.


In [ ]:
# Find the table containing the metrics
    table = soup.find("table", {"id": "metrics-table"})

    # Extract the data from the table
    rows = table.find_all("tr")
    metrics = {}
    for row in rows:
        cols = row.find_all("td")
        if len(cols) == 2:
            metric_name = cols[0].text.strip()
            metric_value = cols[1].text.strip()
            metrics[metric_name] = metric_value

    # Print the metrics
    print(metrics)
else:
    print("Request failed with status code:", response.status_code)

## SETTING THE PERIMETERS

### US Jurisdictions Within 50K-100K Population Range 

In [20]:
import pandas as pd

# Read in the CSV file containing the city population data
city_data = pd.read_csv('/Users/krusty/Downloads/sub-est2021_all.csv', encoding='ISO-8859-1')
#https://www2.census.gov/programs-surveys/popest/datasets/2020-2021/cities/totals/

# Define the lower and upper bounds for the population range you want to select
pop_min = 500000
pop_max = 1000000

# Filter the city data to select only those with population within the specified range
selected_cities = city_data[(city_data['ESTIMATESBASE2020'] >= pop_min) & (city_data['ESTIMATESBASE2020'] <= pop_max)]

# Print the selected cities

print(selected_cities)
len(selected_cities)



       SUMLEV  STATE  COUNTY  PLACE  COUSUB  CONCIT  PRIMGEO_FLAG FUNCSTAT  \
794        50      1      73      0       0       0             0        A   
1111       40      2       0      0       0       0             0        A   
1506      162      4       0  46000       0       0             0        A   
1543      162      4       0  77000       0       0             0        A   
1611      157      4      13  46000       0       0             1        A   
...       ...    ...     ...    ...     ...     ...           ...      ...   
79589      50     55      79      0       0       0             0        A   
79599     157     55      79  53000       0       0             0        A   
79627      61     55      79      0   53000       0             0        F   
79628      71     55      79  53000   53000       0             1        A   
81170      40     56       0      0       0       0             0        A   

                       NAME     STNAME  ESTIMATESBASE2020  POPE

['SUMLEV',
 'STATE',
 'COUNTY',
 'PLACE',
 'COUSUB',
 'CONCIT',
 'PRIMGEO_FLAG',
 'FUNCSTAT',
 'NAME',
 'STNAME',
 'ESTIMATESBASE2020',
 'POPESTIMATE2020',
 'POPESTIMATE2021']

In [53]:
import plotly.graph_objects as go

import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/2014_us_cities.csv')
df.head()

#https://plotly.com/python/bubble-maps/#bubble-map-with-goscattergeo

df = df[(df['pop'] >= 50000) & (df['pop'] <= 100000)] 
df['text'] = df['name'] + '<br>Population ' + (df['pop']/1e6).astype(str)+' million'
limits = [(0,2),(3,10),(11,20),(21,50),(50,3000)]
colors = ["royalblue","crimson","lightseagreen","orange","lightgrey"]
cities = []
scale = 5000

fig = go.Figure()

for i in range(len(limits)):
    lim = limits[i]
    df_sub = df[lim[0]:lim[1]]
    fig.add_trace(go.Scattergeo(
        locationmode = 'USA-states',
        lon = df_sub['lon'],
        lat = df_sub['lat'],
        text = df_sub['text'],
        marker = dict(
            size = df_sub['pop']/scale,
            color = colors[i],
            line_color='rgb(40,40,40)',
            line_width=0.5,
            sizemode = 'area'
        ),
        name = '{0} - {1}'.format(lim[0],lim[1])))

fig.update_layout(
        title_text = '2014 US City Populations Between 50,000 and 100,000<br>(Click legend to toggle traces)',
        showlegend = True,
        geo = dict(
            scope = 'usa',
            landcolor = 'rgb(217, 217, 217)',
        )
    )

fig.show()

In [54]:
import plotly.graph_objects as go
import pandas as pd

#https://simplemaps.com/data/us-cities

df = pd.read_csv('/Users/krusty/Downloads/simplemaps_uscities_basicv1.76/uscities.csv')
df.head()

df = df[(df['population'] >= 50000) & (df['population'] <= 100000)]  # Filter data for populations between 50K and 100K
df['text'] = df['city'] + '<br>population ' + (df['population']/1e6).astype(str)+'million'
limits = [(0,2),(3,10),(11,20),(21,50),(50,3000)]
colors = ["royalblue","crimson","lightseagreen","orange","lightgrey"]
cities = []
scale = 5000

fig = go.Figure()

for i in range(len(limits)):
    lim = limits[i]
    df_sub = df[lim[0]:lim[1]]
    fig.add_trace(go.Scattergeo(
        locationmode = 'USA-states',
        lon = df_sub['lng'],
        lat = df_sub['lat'],
        text = df_sub['text'],
        marker = dict(
            size = df_sub['population']/scale,
            color = colors[i],
            line_color='rgb(40,40,40)',
            line_width=0.5,
            sizemode = 'area'
        ),
        name = '{0} - {1}'.format(lim[0],lim[1])))

fig.update_layout(
        title_text = '2020 US City Populations Between 50,000 and 100,000<br>(Click legend to toggle traces)',
        showlegend = True,
        geo = dict(
            scope = 'usa',
            landcolor = 'rgb(217, 217, 217)',
        )
    )

fig.show()


In [42]:
import plotly.graph_objects as go

import pandas as pd

df = pd.read_csv('/Users/krusty/Downloads/simplemaps_uscities_basicv1.76/uscities.csv')
df.head()

df['text'] = df['city'] + '<br>population ' + (df['population']/1e6).astype(str)+' thousand'
limits = [(5,6),(61,70),(71,80),(81,90),(91,100)]
colors = ["royalblue","crimson","lightseagreen","orange","lightgrey"]
cities = []
scale = 5000

fig = go.Figure()

for i in range(len(limits)):
    lim = limits[i]
    df_sub = df[lim[0]:lim[1]]
    fig.add_trace(go.Scattergeo(
        locationmode = 'USA-states',
        lon = df_sub['lng'],
        lat = df_sub['lat'],
        text = df_sub['text'],
        marker = dict(
            size = df_sub['population']/scale,
            color = colors[i],
            line_color='rgb(40,40,40)',
            line_width=0.5,
            sizemode = 'area'
        ),
        name = '{0} - {1}'.format(lim[0],lim[1])))

fig.update_layout(
        title_text = '2020 US city populations<br>(Click legend to toggle traces)',
        showlegend = True,
        geo = dict(
            scope = 'usa',
            landcolor = 'rgb(217, 217, 217)',
        )
    )

fig.show()

## Heatmap for Similar Population Cities 

In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Read in population data from a CSV file
population_data = pd.read_csv('/Users/krusty/Desktop/Capstone/NRPADataAll.csv')

# Create a pivot table with the population metric as the values and the geographic areas as the rows/columns
pivot_table = population_data.pivot(index='Geographic Area', columns='Year', values='Population Metric')

# Create the heatmap using Seaborn's heatmap function
sns.heatmap(pivot_table, cmap='YlOrRd')

# Add a title and axis labels
plt.title('Population Metric Heatmap')
plt.xlabel('Year')
plt.ylabel('Geographic Area')

# Display the heatmap
plt.show()


KeyError: 'Geographic Area'

## Code for Regression Analysis

In [32]:
# Import the necessary libraries
import pandas as pd
import statsmodels.formula.api as smf

# Read in the data
parks_data = pd.read_csv("/Users/krusty/Desktop/Capstone/NRPADataAll.csv")

# Check the structure of the data
print(parks_data.info())

# Perform the regression analysis
model = smf.ols("num_visitors ~ park_size + num_facilities + maintenance_budget", data=parks_data).fit()

# Check the summary of the model
print(model.summary())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7947 entries, 0 to 7946
Columns: 245 entries, Agency to Have hiring practices and policies that promote a diverse agency workforce?
dtypes: float64(38), int64(1), object(206)
memory usage: 14.9+ MB
None


PatsyError: Error evaluating factor: NameError: name 'num_facilities' is not defined
    num_visitors ~ park_size + num_facilities + maintenance_budget
                               ^^^^^^^^^^^^^^

Code Explanation: 

The smf.ols() function in Python's statsmodels package is used to perform Ordinary Least Squares (OLS) regression analysis.

In the code smf.ols("num_visitors ~ park_size + num_facilities + maintenance_budget", data=parks_data), we are specifying the regression formula. The formula specifies that the dependent variable num_visitors is regressed on the independent variables park_size, num_facilities, and maintenance_budget. The ~ sign in the formula separates the left-hand side (dependent variable) from the right-hand side (independent variables), and the + sign is used to separate the independent variables.

The data parameter is used to specify the dataset to use in the regression analysis. In this case, parks_data is the DataFrame that contains the data on the parks and recs metrics.

The fit() method is used to fit the regression model to the data. Once the model has been fit, the resulting object model contains information about the regression coefficients, standard errors, t-statistics, and p-values, among other things.

Therefore, the entire code model = smf.ols("num_visitors ~ park_size + num_facilities + maintenance_budget", data=parks_data).fit() is creating a linear regression model where num_visitors is the dependent variable and park_size, num_facilities, and maintenance_budget are the independent variables. It then fits the model to the parks_data dataset and stores the results in the model object.